Chat Definitions (#80)
waku-org/specs: informational/chatdefs.md

Establishing a shared definitions document provides several benefits:

1. **Semantic Consistency**: Ensures uniform interpretation of common terms across multiple specifications
2. **Implementation Clarity**: Reduces ambiguity for protocol implementers
3. **Specification Quality**: Enables more precise and concise specification language through reference to established definitions
4. **Ecosystem Coherence**: Promotes consistent understanding within the chat protocol ecosystem.

This specification provides normative definitions that other chat protocol specifications MAY reference to establish precise semantic meaning for common terminology.

## Theory / Semantics

### Definition Categories

Terms are organized into the following categories for clarity and ease of reference:

- **Roles**: Defined entity types that determine how a participant behaves within a communication protocol.
- **Message Types**: Classifications and categories of protocol messages
- **Transports**: Abstract methods of transmitting payloads

## Definitions

> ### Accounts + Identity

Author: Accounts and Identity have yet to be defined for chat Protocols. 

The empty heading was a warning that this work has not been completed. Happy to remove

### Roles

**Sender**: A client which is pushing a payload on to the network, to one or more recipients.

**Recipient**: A client which is the intended receiver of a payload. In a group context there maybe multiple recipients 

**Participant**: A generic term for the rightful members of a conversation. Senders and Recipients are roles that participants can hold.

### Message Types

The term "message" often has multiple meanings depending on context. The following definitions are used to disambiguate between different abstraction levels. 

**Content**: The data that is intended for end-users or applications - this includes text, images and other user valued data. These messages originate from the application, and are ultimately delivered to other applications. 

> **Frame**: The data generated by a chat protocol. Frames are messages used by protocols to provide functionality - they are how clients exchange information with each other and coordinate state.

Author: Updated

> **Payload**: The encoded data or data structures used by lower layer protocols. This includes the encoded bytes passed to a delivery service or subsequent layers. Payloads are usually associated with the delivery process, which is not defined by chat protocols.

Reviewer: I think it would be nice to describe how Content/Frame/Payload relate to each other

> **Payload**: The encoded data or data structures used by lower layer protocols. This includes the encoded bytes passed to a delivery service or subsequent layers. Payloads are usually associated with the delivery process, which is not defined by chat protocols.

Author: Updated

Other specific message types include:

> **Content Type**: The structured format of content. These data structures represent specific encodings of Text, Image, Audio data. 

Reviewer 2: > **Content Type**: The structured format of content.

That reads a bit oddly to me. I usually think of Content Type as a label or classification, not the data structure itself.

```suggestion
**Content Type**: A classification that identifies the kind of Content being transmitted (e.g., Text, Image, Audio). Content Types determine how the associated data is structured and interpreted by applications.
```

Then I am not sure whether it fits under "Other specific message types include".

Author: This is my latest take on it. Any better?

```suggestion
**Content Type**: A definition of the structure and encoding of a Content instance, interpreted solely by the Application.
```

> **Delivery Acknowledgement**: A notification from a receiving client to sender that their message was successfully received. While similar to a read-receipt delivery acknowledgements differ in that the acknowledgement originates based on the client, where read-receipts are fired when they are displayed to a user.

Reviewer: ```suggestion
**Delivery Acknowledgement**: A notification from a receiving client to sender that their message was successfully received. While similar to a read-receipt, delivery acknowledgements differ in that the acknowledgement originates based on the client, where read-receipts are fired when they are displayed to a user.
```

> **Delivery Acknowledgement**: A notification from a receiving client to sender that their message was successfully received. While similar to a read-receipt delivery acknowledgements differ in that the acknowledgement originates based on the client, where read-receipts are fired when they are displayed to a user.

Reviewer 2: `Invite` explicitly specifies that it is a frame. Should the delivery ack also be a frame?

```suggestion
**Delivery Acknowledgement**: A notification frame from a receiving client to sender that their message was successfully received. While similar to a read-receipt delivery acknowledgements differ in that the acknowledgement originates based on the client, where read-receipts are fired when they are displayed to a user.
```

4. **Ecosystem Coherence**: Promotes consistent understanding within the chat protocol ecosystem.

This specification provides normative definitions that other chat protocol specifications MAY reference to establish precise semantic meaning for common terminology.

## Theory / Semantics

### Definition Categories

Terms are organized into the following categories for clarity and ease of reference:

- **Roles**: Defined entity types that determine how a participant behaves within a communication protocol.
- **Message Types**: Classifications and categories of protocol messages
- **Transports**: Abstract methods of transmitting payloads
## Definitions

### Accounts + Identity

### Roles

**Sender**: A client which is pushing a payload on to the network, to one or more recipients.

**Recipient**: A client which is the intended receiver of a payload. In a group context there maybe multiple recipients 

**Participant**: A generic term for the rightful members of a conversation. Senders and Recipients are roles that participants can hold.

### Message Types

**Payload**: The encoded bytes as produced by a chat protocol. The term `message` is avoided due to conflicts with other layers.

> **Frame**: A data structure which is required to implement a chat protocol. Frames are used to disambiguate the objects defined by chat protocols from transport 'messages' and application `content`. 

Author: Absolutely - Examples though are provided in the conversation specifications. This definitions spec is intended to only define the terms used to provide consistent vocabulary. Frames definitions are arbitrary, and defined by their use case. The different in the 3 terms is the intent and purpose of the "Message".

> **Frame**: A data structure which is required to implement a chat protocol. Frames are used to disambiguate the objects defined by chat protocols from transport 'messages' and application `content`. 

Author: Updated

**Content**: The data that is intended for end-users or applications. Chat protocols transmit and route content between clients. 

**Content Type**: The structured format of content. These data structures represent specific encodings of Text, Image, Audio data. 

**Delivery Acknowledgement**: A notification from a receiving client to sender that their message was successfully received. While similar to a read-receipt delivery acknowledgements differ in that the acknowledgement originates based on the client, where read-receipts are fired when they are displayed to a user.

**Invite**: An protocol message from one client to another to establish a communication channel. Invites notify a client that someone wants to communicate with them, and provides the required information to do so.

### Transports

> **Out-of-Band**: The transfer of information using a channel separate from the defined protocol channel. Data sent Out-of-Band is transmitted using an undefined mechanism. This is used when protocols requires information to be shared with another entity, but it does not describe how that should occur. The responsibility to define how this occurs is the implementer or other protocols in the suite. 

Reviewer 3: ```suggestion
**Out-of-Band**: The transfer of information using a channel separate from the defined chat protocol. Data sent Out-of-Band is transmitted using an undefined mechanism. This is used when protocols requires information to be shared with another entity, but it does not describe how that should occur. The responsibility to define how this occurs is the implementer or other protocols in the suite. 
```

### Software Entities

> **Client**: Software that implements a chat protocol. It is responsible for establishing and maintaining communication with the underlying network, performing protocol operations (e.g., encryption, key management, frame generation), and exposing an interface for higher-level software. A Client serves as the technical component that provides messaging capabilities to Applications.

Reviewer 3: The term client seems not clear, maybe just use chat-sdk implementation here.
